pkg/gitproviders/provider_user.go

```diff
@@ -0,0 +1,225 @@
+func (p userGitProvider) UploadDeployKey(owner, repoName string, deployKey []byte) error {
+	deployKeyInfo := gitprovider.DeployKeyInfo{
+		Name:     deployKeyName,
+		Key:      deployKey,
+		ReadOnly: gitprovider.BoolVar(false),
+	}
+
+	ctx := context.Background()
+	defer ctx.Done()
+
+	userRef := NewUserRepositoryRef(p.domain, owner, repoName)
+	userRepo, err := p.provider.UserRepositories().Get(ctx, userRef)
+
+	if err != nil {
+		return fmt.Errorf("error getting user repo reference for owner %s, repo %s, %s ", owner, repoName, err)
+	}
+
+	fmt.Println("uploading deploy key")
+
+	_, err = userRepo.DeployKeys().Create(ctx, deployKeyInfo)
+	if err != nil {
+		return fmt.Errorf("error uploading deploy key %s", err)
+	}
+
+	if err = utils.WaitUntil(os.Stdout, time.Second, defaultTimeout, func() error {
+		_, err = userRepo.DeployKeys().Get(ctx, deployKeyName)
+		return err
+	}); err != nil {
+		return fmt.Errorf("error verifying deploy key %s existance for repo %s. %s", deployKeyName, repoName, err)
+	}
+
+	return nil
+}
+
+func (p userGitProvider) GetDefaultBranch(url string) (string, error) {
+	repoInfoRef, err := p.getRepoInfoFromUrl(url)
+	if err != nil {
+		return "main", err
+	}
+
+	return *repoInfoRef.DefaultBranch, nil
+}
+
+func (p userGitProvider) GetRepoVisibility(url string) (*gitprovider.RepositoryVisibility, error) {
+	repoInfoRef, err := p.getRepoInfoFromUrl(url)
+	if err != nil {
+		return nil, err
+	}
+
+	return repoInfoRef.Visibility, nil
+}
+
+func (p userGitProvider) getRepoInfoFromUrl(repoUrl string) (*gitprovider.RepositoryInfo, error) {
+	normalizedUrl, err := NewNormalizedRepoURL(repoUrl)
+	if err != nil {
+		return nil, err
+	}
+
+	repo, err := p.getUserRepo(normalizedUrl.Owner(), normalizedUrl.RepositoryName())
+	if err != nil {
+		return nil, err
+	}
+
+	repoInfo := repo.Get()
+
+	return &repoInfo, nil
+}
+
+func (p userGitProvider) getUserRepo(user string, repoName string) (gitprovider.UserRepository, error) {
+	ctx := context.Background()
+	defer ctx.Done()
+
+	userRepoRef := NewUserRepositoryRef(p.domain, user, repoName)
+
+	repo, err := p.provider.UserRepositories().Get(ctx, userRepoRef)
+	if err != nil {
+		return nil, fmt.Errorf("error getting user repository %w", err)
+	}
+
+	return repo, nil
+}
+
+func (p userGitProvider) CreatePullRequest(owner string, repoName string, targetBranch string, newBranch string, files []gitprovider.CommitFile, commitMsg string, prTitle string, prDescription string) (gitprovider.PullRequest, error) {
+	pr, err := p.createPullRequestToUserRepo(owner, repoName, targetBranch, newBranch, files, commitMsg, prTitle, prDescription)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create pull request: %w", err)
+	}
+
+	return pr, nil
+}
+
+func (p userGitProvider) createPullRequestToUserRepo(owner string, repoName string, targetBranch string, newBranch string, files []gitprovider.CommitFile, commitMessage string, prTitle string, prDescription string) (gitprovider.PullRequest, error) {
```

Comment:
Does it make sense a good portion of these parameters into an object like `PullRequestInfo`?

```
type PullRequestInfo struct {
  title string,
  description string,
  commitMessage string,
  files []gitprovider.CommitFile
}

These always seem like they would go together.

Reply:
💯